components/TextFields/src/MDCTextInputBehavior.h

```diff
@@ -0,0 +1,163 @@
+ limitations under the License.
+ */
+
+#import <UIKit/UIKit.h>
+
+@protocol MDCTextInput;
+@protocol MDCTextInputCharacterCounter;
+
+/**
+ Presentation styles for a text input. The style determines specific aspects of the text
+ input, such as sizing, placeholder placement and behavior, layout, etc.
+ */
+typedef NS_ENUM(NSUInteger, MDCTextInputPresentationStyle) {
+  /**
+   Default style with an inline placeholder (that disappears when text is entered) and character
+   count / limit below text.
+   */
+  MDCTextInputPresentationStyleDefault = 0,
+
+  /**
+   The placeholder text is laid out inline but will float above the field when there is content or
+   the field is being edited. The character count is below text. The Material Design guidelines
+   call this 'Floating inline labels.'
+   https://material.io/guidelines/components/text-fields.html#text-fields-labels
+   */
+  MDCTextInputPresentationStyleFloatingPlaceholder,
+
+  /** The placeholder is laid out inline and the character count is also inline to the right. */
+  MDCTextInputPresentationStyleFullWidth,
+};
+
+/**
+ Material Design themed states for textInputs. The logic for 'automagic' error states changes:
+ underline color, underline text color.
+ https://www.google.com/design/spec/components/text-fields.html#text-fields-single-line-text-field
+ */
+@interface MDCTextInputBehavior : NSObject <NSCoding, NSCopying>
+
+/** The text input the behavior is managing. */
+@property(nonatomic, nullable, weak) UIView<MDCTextInput> *textInput NS_SWIFT_NAME(input);
+
+/**
+ The color used to denote error state in the underline, the errorText's label, the plceholder and
+ the character count label.
+
+ Default is red.
+ */
+@property(nonatomic, nullable, strong) UIColor *errorColor UI_APPEARANCE_SELECTOR;
+
+/**
+ Controls when the underline will be shown.
+
+ The underline is an overlay that can be hidden depending on the editing state of the input text.
+
+ Default is UITextFieldViewModeAlways.
+ */
+@property(nonatomic, assign) UITextFieldViewMode underlineViewMode NS_SWIFT_NAME(underlineMode)
+    UI_APPEARANCE_SELECTOR;
+
+/**
+ Optional, custom character counter.
+
+ Instead of relying on the default character count which is naive (counts each character regardless
+ of context), this object can instead choose to do sophisticated counting (ie: ignoring whitespace,
+ ignoring url strings, etc).
+ */
+@property(nonatomic, nullable, weak) IBInspectable id<MDCTextInputCharacterCounter>
+    characterCounter;
+
+/**
+ Controls when the character count will be shown.
+
+ Default is UITextFieldViewModeNever.
+ */
+@property(nonatomic, assign) UITextFieldViewMode characterCountViewMode NS_SWIFT_NAME(characterMode)
+    ;
+
+/**
+ The character limit for the text input. A label under the input counts characters entered and
+ presents the count / the limit.
+
+ Default is 0.
+ */
+@property(nonatomic, assign) IBInspectable NSUInteger characterLimit;
```

Comment:
how does this interplay with the setting the setErrorText: APIs? which one wins. How does a client see the results of the out of bounds case?